src/hooks/use-transition-component.tsx

```diff
@@ -0,0 +1,104 @@
+import type { ComponentChildren } from 'preact';
+import type { JSX } from 'preact';
+import { useCallback, useEffect, useState } from 'preact/hooks';
+
+import type { TransitionComponent } from '../types';
+
+type TransitionComponentOptions = {
+  closed: boolean;
+  transitionComponent?: TransitionComponent;
+  onClose?: () => void;
+  onOpen?: () => void;
+};
+
+type TransitionComponentResult = {
+  closeHandler: () => void;
+  isClosed: boolean;
+  wrapWithTransition: (children: ComponentChildren) => JSX.Element;
+};
+
+export function useTransitionComponent({
+  closed,
+  transitionComponent,
+  onClose,
+  onOpen,
+}: TransitionComponentOptions): TransitionComponentResult {
+  // TODO To properly handle closing/opening with a TransitionComponent, these
+  //  two pieces of state need to be synced with the `closed` argument.
+  //  That makes the logic hard to follow. It would be good to revisit eventually
+  const [isClosed, setIsClosed] = useState(closed);
+  const [transitionComponentVisible, setTransitionComponentVisible] =
+    useState(false);
+
+  const closeHandler = useCallback(() => {
+    if (transitionComponent) {
+      // When a TransitionComponent is provided, the actual "onClose" will be
+      // called by that component, once the "out" transition has finished
+      setTransitionComponentVisible(false);
+    } else {
+      onClose?.();
+    }
+  }, [onClose, transitionComponent]);
+
+  const onTransitionEnd = useCallback(
+    (direction: 'in' | 'out') => {
+      if (direction === 'in') {
+        onOpen?.();
+      } else {
+        setIsClosed(true);
+        onClose?.();
+      }
+    },
+    [onOpen, onClose]
+  );
+
+  useEffect(() => {
+    if (closed) {
+      if (transitionComponent) {
+        setTransitionComponentVisible(false);
+      } else {
+        setIsClosed(true);
+      }
+    } else {
+      setIsClosed(false);
+    }
+
+    if (!closed && !transitionComponent) {
+      onOpen?.();
+    }
+
+    // We only want to run this effect when opened or closed.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [closed]);
+
+  useEffect(() => {
+    if (!isClosed) {
+      setTransitionComponentVisible(true);
+    }
+  }, [isClosed]);
+
+  const wrapWithTransition = useCallback(
+    (children: ComponentChildren) => {
+      if (!transitionComponent) {
+        return <>{children}</>;
+      }
+
+      const TransitionComp = transitionComponent;
+      return (
+        <TransitionComp
+          direction={transitionComponentVisible ? 'in' : 'out'}
+          onTransitionEnd={onTransitionEnd}
+        >
+          {children}
+        </TransitionComp>
+      );
+    },
+    [transitionComponent, transitionComponentVisible, onTransitionEnd]
+  );
```

Comment:
My initial idea to hide the logic of conditionally wrapping the consuming component in a `TransitionComponent` or a `Fragment` was to expose an actual component:

```tsx
const Wrapper = ({ children }) => {
  if (!transitionComponent) {
    return <>{children}</>;
  }

  const TransitionComp = transitionComponent;
  return <TransitionComp ... />;
};
```

And then use it like this in the consuming component:

```tsx
const { Wrapper, ... } = useTransitionComponent(...);

return (
  <Wrapper>
    <div>...</div>
  </Wrapper>
);
```

However, that introduces a component which is defined within a component, and the effects didn't work as expected.

That's why I ended up creating a callback that does the wrapping.